### ISSUE TITLE
DNR

### ISSUE THREAD

**Issue by pierreozoux** [indiehosters/documents#17]
Problem:
When a user register, he needs to choose a Domain name.

we need a nice user interface to guide the user thought this process.

Once he decides, the information should be passed to the CRM, that will actually order the domain name in the bacground.

I'd love to hear feedback from cozy, yunohost, reclaim, goclouder on how they solve or plan to solve that issue. And also, how we could collaborate on some pieces.

**Comment by michielbdejong** [comment]
About the "CRM, that will actually order the domain name" part, from https://github.com/indiehosters/documents/issues/15#issuecomment-93339198:
> I realized the webhosting automation is missing from Odoo. If you look at https://opensrs.com/integration/tools/ you would want at least one open source tool to be in that list, but there is none! Looking also at http://alternativeto.net/software/whmcs/?license=free and http://stackoverflow.com/questions/29315895/best-whmcs-open-source-alternative the closest thing is https://github.com/TheHostingTool/TheHostingTool but it seems to me we should go for Odoo and add modules for webhosting-automation to it.

About the "nice user interface", I think most shared webhosting signup pipelines are way too overwhelming. I would put the payment before the choice of domain name (that's the way we have it now, with the 'Subscribe and you will be contacted by your hoster' button).

Once we launch free *.indie.host hosting, I think signing up for a domain name without signing up for freebie first is going to be a rare case, which we don't need to tailor for. So then we could make it "switch to your own identity (I have a domain name)" and  "switch to your own identity (choose a domain name)"

**Comment by timmmmyboy** [comment]
We're using WHMCS but certainly have dreams of moving away from it (proprietary is one thing, but they also encrypt all their code so I can't even make changes). We hit the same wall of not finding a whole lot of open source stuff out there for web hosting, it's a sad state of the industry. Part of why I think the onboarding process is overwhelming at a lot of hosts is that in addition to the industry norm of attempting to upsell the user during the checkout process, there's also a lot of information to be gathered (contact info, billing info, domain/hosting info, etc).

The domain choice piece is interesting because on one hand it can paralyze a user that hasn't thought about that yet or is unsure. In those cases maybe doing all the billing first is ideal (although seems like 6 in one hand, half dozen in the other, they still have to go through the process regardless of what order they do it). On the other hand some folks enjoy the process of finding a name they can call their own and in fact that's what they feel they're "buying" with an account so making them checkout first is going to feel less personalized since they're paying for something generic and not unique. Just some thoughts.

**Comment by michielbdejong** [comment]
@timmmmyboy would you consider switching from WHMCS to Odoo as a CRM if we develop all the necessary modules for triggering a full hosting product deploy? Can you list the things WHMCS currently does for you? It adds domains into cPanel, right?

**Comment by timmmmyboy** [comment]
@michielbdejong Quite possibly, I'd have to play around with their system. I'm assuming it functions something like a plugin ecosystem, surely all the apps they list on their site aren't a part of the install? That would be a nightmare. I do like the idea of selecting some core components to put together from what they have but I'd have to see what a migration process would look like too.

As far as what we use WHMCS for:
* Domain Registration (3 different registrar modules interfacing with the API for the registrars we use)
* Payment Gateway (Paypal and Braintree modules)
* cPanel Provisioning (cPanel module)
* Some notifications/messaging, for example I had to send a message to all users on a particular server that had active accounts and that was the easiest way to do it through WHMCS. Reminder emails and renewals and all that
* Invoicing and Quote generation
* Fraud monitoring (via Maxmind)
* Affiliate system (although I found there's pretty lackluster so I'm not married to it) - We mostly use this to allow folks to generate a coupon they can give students.

We've done a handful of customizations, mostly to the theme on the frontend, to integrate and extend it in various ways as well. One thing that has been a security blanket for me is that WHMCS is PHP which I'm very comfortable working with. Looks like Odoo is python so I'd want to look at what developing an "app" for it and modifying existing ones requires.

**Comment by michielbdejong** [comment]
Thanks! Yeah, I'm also more comfortable with php than with python. One reason to go with Odoo is that they seem to be quite big (even though their target is ERP for any business, not specifically the hosting business), and https://github.com/clouder-community/clouder are using them, so that could lead to collaboration and code reuse.

I will wait what THT say on https://github.com/TheHostingTool/TheHostingTool/issues/144, it's not so clear what the status is, but who knows, maybe there is still some momentum there which we can piggyback on. They are in php, and it seems they are the only open source tool aimed specifically at hosting (the only other one seems to be BoxBilling, which is published under an Apache license, but with a license restriction, so it is not entirely clear whether the author really wants people to jailbreak that limitation and use it without paying).

**Comment by YannickB** [comment]
On Clouder, for now we can only manage subdomain (for example if someone ask me to create a new instance, it create newinstance.clouder.at). You can easily manage your own root domain, but you have to configure it manually, while subdomain creation are automated.

The DNS server is managed inside a container on my servers, so as long as the main domain (here clouder.at) point to my servers everything is automated.
There is currently no way to automatically buy with Clouder (or Odoo) a domain from the registrar, but it shall not be too complicated to do.

**Comment by YannickB** [comment]
Just one word for the marketing positionment of Odoo, they are not calling themselves an ERP anymore (this is why they changed their name from OpenERP), they are a business suite application. I think what we are doing is completely compatible with the their marketing I think.

I'll be making a conference at Odoo Experience in June, I expect many Odoo developers to join us at this moment.

**Comment by sysfu** [comment]
If the goal is to streamline the signup/registration process to require the absolute minimum number of decisions to be made by the user, I think it makes sense to defer the choice of selecting a registered domain name until after the user has signed up.

The problem description includes the statement "When a user registers, he needs to choose a Domain name". Is this assumption necessarily true? Could a username or other unique identifier such as email address be substituted instead? If a domain name is required at signup, maybe a brief description of the technical reasons for doing so could be given in this thread?

The subdomain strategy mentioned by Yannick is used by the popular build your own web site service Yola.com. Say I were to sign up at Yola. First I register an account using an email address. No domain name required. Then via the dashboard I create a web site, sysfu, which is automatically published at subdomain sysfu.yolasite.com, still no domain name required.

If I want to, my new web sites 'lives' there for as long as I want. Maybe I decide I want to kick it up a notch, register my own domain name, and re-publish my pre-existing web site at either sysfu.com or www.sysfu.com. No problem, Yola allows me to do this.

What I envision is a process whereby users can get started by simply registering an account, and then start creating services at their own subdomain(s) without having to bother with *any* of the domain name registration or TLS cert stuff. The only restriction is that the subdomain they choose is not already taken.

This removes a huge psychological and logistical barrier to getting users on board. I can only speak for myself but just the prospect being forced to select a new domain during the signup process would create a lot of anxiety and mental friction.

**Comment by michielbdejong** [comment]
Yes, and I was thinking to allow two-click signup with Twitter, and then you just get `<handle>.indie.host`, no further clicks required. And indeed, the domain name would be an upsell then, for existing users.

**Comment by YannickB** [comment]
I agree with sysfu, in most case it's best to propose just a subdomain at registration, and then either : 
-The customer buy the domain at a registrar and make a CNAME
-We adapt our dev so they can automatically buy their own domain, but it shall be done after they subscribed.
